spring-client-common/src/main/java/io/camunda/commons/auth/SelfManagedAuthentication.java

```diff
@@ -0,0 +1,115 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.UnsupportedEncodingException;
+import java.lang.invoke.MethodHandles;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.AbstractMap;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static io.camunda.commons.http.JsonUtils.toResult;
+
+public class SelfManagedAuthentication extends JwtAuthentication {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  private String authUrl;
+  private String keycloakRealm = "camunda-platform";
+  private String keycloakUrl;
+  private JwtConfig jwtConfig;
+  private Map<Product, String> tokens;
+  private Map<Product, Integer> expirations;
+
+  public SelfManagedAuthentication() {
+    tokens = new HashMap<>();
+    expirations = new HashMap<>();
+  }
+
+  public SelfManagedAuthentication jwtConfig(JwtConfig jwtConfig) {
+    this.jwtConfig = jwtConfig;
+    return this;
+  }
+
+  public SelfManagedAuthentication keycloakUrl(String keycloakUrl) {
+    this.keycloakUrl = keycloakUrl;
+    return this;
+  }
+
+  public SelfManagedAuthentication keycloakRealm(String keycloakRealm) {
+    if (keycloakRealm != null) {
+      this.keycloakRealm = keycloakRealm;
+    }
+    return this;
+  }
+
+  @Override
+  public Authentication build() {
+    authUrl = keycloakUrl+"/auth/realms/"+keycloakRealm+"/protocol/openid-connect/token";
+    jwtConfig.getMap().forEach(this::retrieveToken);
+    return this;
+  }
+
+  private void retrieveToken(Product product, Credential credential) {
+    try {
+      HttpPost httpPost = new HttpPost(authUrl);
+      httpPost.addHeader("Content-Type", "application/x-www-form-urlencoded");
+
+      Map<String, String> parameters = new HashMap<>();
+      parameters.put("grant_type", "client_credentials");
+      parameters.put("client_id", credential.clientId);
+      parameters.put("client_secret", credential.clientSecret);
+
+      String form = parameters.entrySet()
+        .stream()
+        .map(e -> {
+          try {
+            return e.getKey() + "=" + URLEncoder.encode(e.getValue(), StandardCharsets.UTF_8.toString());
+          } catch (UnsupportedEncodingException ex) {
+            throw new RuntimeException(ex);
+          }
+        })
+        .collect(Collectors.joining("&"));
+
+      httpPost.setEntity(new StringEntity(form));
+      CloseableHttpClient client = HttpClients.createDefault();
+      CloseableHttpResponse response = client.execute(httpPost);
+      TokenResponse tokenResponse = toResult(EntityUtils.toString(response.getEntity()), TokenResponse.class);
+      // TODO: verify JWT has the desired permission vs what the user requested
+      tokens.put(product, tokenResponse.getAccessToken());
+      expirations.put(product, (int) (System.currentTimeMillis()/1000 + tokenResponse.getExpiresIn()));
```

Comment:
Why do we convert it to seconds when we have to convert it again during refresh? Maybe just store it in `ms`.

Reply:
Switched to `LocalDateTime` and got rid of seconds / milliseconds in https://github.com/camunda-community-hub/spring-zeebe/pull/510/commits/0b537aa7a8ff6903bea1a85a915e2f184495f95d